### ISSUE TITLE
Healthy Agent suddenly stopped working 

### ISSUE THREAD

**Issue by Erick-Reyes** [gravitational/teleport#12527]
Expected behavior:
Healthy agent to continue to work properly and visible when outputting ```tsh ls``` or ```tctl get node```.

Current behavior:
Basically, a healthy agent just seems to stop working. It was an 8.3.4 agent connected to a teleport cloud backend. This agent had been running for 4 weeks total. The last time it had output anything (as seen in the logs in journalctl) was on march 28th. The node no longer appeared in tsh ls or tctl get node. The teleport cloud tenant it was connected to was at version 8.3.7.
We grabbed a stacktrace by sending in a USR1 signal before restarting the agent; this is attach to the bug. 

```
root@example:~# journalctl -fu teleport
-- Logs begin at Tue 2022-03-22 11:19:40 UTC. --
Mar 25 13:47:41 example.local teleport[904]: 2022-03-25T13:47:41Z INFO [NODE:PROX] Connected. addr:172.21.74.5x:47186 remote-addr:35.71.146.110:11261 leaseID:346 target:example.teleport.sh:11261 reversetunnel/agent.go:403
Mar 25 13:47:42 example.local teleport[904]: 2022-03-25T13:47:42Z INFO [NODE:PROX] Connected. addr:172.21.74.5x:44460 remote-addr:52.223.11.70:11261 leaseID:347 target:example.teleport.sh:11261 reversetunnel/agent.go:403
Mar 25 13:47:44 example.local teleport[904]: 2022-03-25T13:47:44Z INFO [NODE:PROX] Connected. addr:172.21.74.5x:47188 remote-addr:35.71.146.110:11261 leaseID:348 target:example.teleport.sh:11261 reversetunnel/agent.go:403
Mar 28 09:12:07 example.local teleport[904]: 2022-03-28T09:12:07Z INFO [NODE:PROX] Connection closed, returning leaseID:264 target:example.teleport.sh:11261 reversetunnel/agent.go:551
Mar 28 09:12:07 example.local teleport[904]: 2022-03-28T09:12:07Z WARN [NODE:PROX] Failed to close discovery channel:: EOF leaseID:264 target:example.teleport.sh:11261 reversetunnel/agent.go:540
Mar 28 09:12:07 example.local teleport[904]: 2022-03-28T09:12:07Z WARN [NODE:PROX] Unable to continue processioning requests: heartbeat: connection closed. leaseID:264 target:example.teleport.sh:11261 reversetunnel/agent.go:427
Mar 28 09:12:07 example.local teleport[904]: 2022-03-28T09:12:07Z WARN [NODE:PROX] Failed to close remote tunnel: close tcp 172.21.74.5x:47054->35.71.146.110:11261: use of closed network connection, local addr: 172.21.74.5x:47054 remote addr: 35.71.146.11x:11261 leaseID:264 target:example.teleport.sh:11261 reversetunnel/agent.go:395
Mar 28 09:12:09 example.local teleport[904]: 2022-03-28T09:12:09Z INFO [NODE:PROX] Connected. addr:172.21.74.5x:44504 remote-addr:52.223.11.7x:11261 leaseID:349 target:example.teleport.sh:11261 reversetunnel/agent.go:403
Mar 28 09:12:10 example.local teleport[904]: 2022-03-28T09:12:10Z INFO [NODE:PROX] Connected. addr:172.21.74.5x:47192 remote-addr:35.71.146.11x:11261 leaseID:350 target:example.teleport.sh:11261 reversetunnel/agent.go:403
Mar 28 09:12:11 example.local teleport[904]: 2022-03-28T09:12:11Z INFO [NODE:PROX] Connected. addr:172.21.74.5x:44508 remote-addr:52.223.11.7x:11261 leaseID:351 target:example.teleport.sh:11261 reversetunnel/agent.go:403
```

The interesting item is the agent reconnects just fine when the process is restarted. No upgrade/changing happening at this time. 

Our ask: 
We're asking devs for input because we're not sure what could cause an agent to be unresponsive for that long.
This is just an agent running via systemd, and our systemd unit doesn't ship with any healthchecks, and our default config doesn't enable 0.0.0.0:3000 so the readyz endpoint isn't available.
Additionally, I've seen the readyz endpoint return healthy in similar agent situations where there's no tunnel connection and the service isn't actually available.
This behavior seems to be quite similar to the 8.2.0 bug where an agent would fail to reestablish a tunnel. This agent had incremented its lease count, which suggests that it was doing some reconnecting for some reason. I'd still expect it to retry until it reconnects or at least prints info in the logs if it can't get a tunnel established.
My fear is that this is another trigger for the same behavior that 8.2.0 had, but perhaps it happens less frequently or only happens in a very specific circumstance.
Even if /readyz correctly reflected the non-ready state, a bug that leads to this situation is a potentially high impact since teleport agents can simply disappear without warning from the point of view of the user.

Maybe a good action item might be to add some more logging to the code that tries to keep the tunnel open? Maybe we could implement the ability for teleport to use the systemd watchdog service for systemd deployments?

Bug details:
- Teleport version: v.8.3.4
- Recreation steps: Deploy node with v8.3.4 and issue re-occurs without any major errors found when putting servers in debug mode
- Debug logs: Attach to issue 
[stacktrace_4859.log](https://github.com/gravitational/teleport/files/8654842/stacktrace_4859.log)

**Comment by Erick-Reyes** [comment]
Issue re-occurred again 5/9/22:

root@example:~# cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.2 LTS"

root@example:~# dpkg -l |grep -i teleport
ii  teleport                           8.3.8                             amd64        Teleport is a gateway for managing access to clusters of Linux servers via SSH or the Kubernetes API [64-bit x86 Open source edition]

root@example:~# journalctl -fu teleport
-- Logs begin at Sun 2022-03-27 02:12:12 UTC. --
May 01 13:58:27 example.local teleport[3450769]: 2022-05-01T13:58:27Z INFO [NODE:PROX] Connection closed, returning leaseID:21 target:example.teleport.sh:11261 reversetunnel/agent.go:557
May 01 13:58:27 example.local teleport[3450769]: 2022-05-01T13:58:27Z WARN [NODE:PROX] Failed to close discovery channel:: EOF leaseID:21 target:example.teleport.sh:11261 reversetunnel/agent.go:546
May 01 13:58:27 example.local teleport[3450769]: 2022-05-01T13:58:27Z WARN [NODE:PROX] Unable to continue processioning requests: heartbeat: connection closed. leaseID:21 target:example.teleport.sh:11261 reversetunnel/agent.go:433
May 01 13:58:27 example.local teleport[3450769]: 2022-05-01T13:58:27Z WARN [NODE:PROX] Failed to close remote tunnel: close tcp 172.21.2.x:41022->35.71.146.11x:11261: use of closed network connection, local addr: 172.21.2.x:41022 remote addr: 35.71.146.11x:11261 leaseID:21 target:example.teleport.sh:11261 reversetunnel/agent.go:401
May 01 13:58:28 example.local teleport[3450769]: 2022-05-01T13:58:28Z INFO [NODE:PROX] Connected. addr:172.21.2.x:40408 remote-addr:35.71.146.11x:11261 leaseID:615 target:example.teleport.sh:11261 reversetunnel/agent.go:409
May 01 13:58:31 example.local teleport[3450769]: 2022-05-01T13:58:31Z INFO [NODE:PROX] Connected. addr:172.21.2.x:39212 remote-addr:52.223.11.7x:11261 leaseID:616 target:example.teleport.sh:11261 reversetunnel/agent.go:409
May 01 13:58:31 example.local teleport[3450769]: 2022-05-01T13:58:31Z INFO [NODE:PROX] Connected. addr:172.21.2.x:39216 remote-addr:52.223.11.7x:11261 leaseID:617 target:example.teleport.sh:11261 reversetunnel/agent.go:409
May 01 13:58:33 example.local teleport[3450769]: 2022-05-01T13:58:33Z INFO [NODE:PROX] Connected. addr:172.21.2.x:40420 remote-addr:35.71.146.11x:11261 leaseID:618 target:example.teleport.sh:11261 reversetunnel/agent.go:409
May 01 13:58:35 example.local teleport[3450769]: 2022-05-01T13:58:35Z INFO [NODE:PROX] Connected. addr:172.21.2.x:39220 remote-addr:52.223.11.7x:11261 leaseID:619 target:example.teleport.sh:11261 reversetunnel/agent.go:409
May 01 13:58:38 example.local teleport[3450769]: 2022-05-01T13:58:38Z INFO [NODE:PROX] Connected. addr:172.21.2.x:39228 remote-addr:52.223.1x.70:11261 leaseID:620 target:example.teleport.sh:11261 reversetunnel/agent.go:409

Server is missing from ```tsh ls``` and as usual restart makes it work again.

**Comment by zmb3** [comment]
Closing due to inactivity. @Erick-Reyes please open a new issue if there are still problems.
